Tools/build/compute-changes.py

```diff
@@ -0,0 +1,195 @@
+
+    write_github_output(outputs)
+
+
+def git_branches() -> tuple[str, str]:
+    target_branch = os.environ.get("GITHUB_BASE_REF", "")
+    target_branch = target_branch.removeprefix("refs/heads/")
+    print(f"target branch: {target_branch!r}")
+
+    head_branch = os.environ.get("GITHUB_HEAD_REF", "")
+    head_branch = head_branch.removeprefix("refs/heads/")
+    print(f"head branch: {head_branch!r}")
+    return target_branch, head_branch
+
+
+def get_changed_files(ref_a: str = "main", ref_b: str = "HEAD") -> Set[Path]:
+    """List the files changed between two Git refs, filtered by change type."""
+    args = ("git", "diff", "--name-only", f"{ref_a}...{ref_b}", "--")
+    print(*args)
+    changed_files_result = subprocess.run(
+        args, stdout=subprocess.PIPE, check=True, encoding="utf-8"
+    )
+    changed_files = changed_files_result.stdout.strip().splitlines()
+    return frozenset(map(Path, filter(None, map(str.strip, changed_files))))
+
+
+def process_changed_files(changed_files: Set[Path]) -> Outputs:
+    run_tests = False
+    run_ci_fuzz = False
+    run_docs = False
+    run_windows_msi = False
+
+    for file in changed_files:
+        # Documentation files
+        doc_or_misc = file.parts[0] in {"Doc", "Misc"}
+        doc_file = file.suffix in SUFFIXES_DOCUMENTATION or doc_or_misc
+
+        if file.parent == GITHUB_WORKFLOWS_PATH:
+            if file.name == "build.yml":
+                run_tests = run_ci_fuzz = True
+            if file.name == "reusable-docs.yml":
+                run_docs = True
+            if file.name == "reusable-windows-msi.yml":
+                run_windows_msi = True
+
+        if not (
+            doc_file
+            or file == GITHUB_CODEOWNERS_PATH
+            or file.name in CONFIGURATION_FILE_NAMES
+        ):
+            run_tests = True
+
+        # The fuzz tests are pretty slow so they are executed only for PRs
+        # changing relevant files.
+        if file.suffix in SUFFIXES_C_OR_CPP:
+            run_ci_fuzz = True
+        if file.parts[:2] in {
+            ("configure",),
+            ("Modules", "_xxtestfuzz"),
+        }:
+            run_ci_fuzz = True
+
+        # Check for changed documentation-related files
+        if doc_file:
+            run_docs = True
+
+        # Check for changed MSI installer-related files
+        if file.parts[:2] == ("Tools", "msi"):
+            run_windows_msi = True
+
+    return Outputs(
+        run_ci_fuzz=run_ci_fuzz,
+        run_docs=run_docs,
+        run_tests=run_tests,
+        run_windows_msi=run_windows_msi,
+    )
+
+
+def process_target_branch(outputs: Outputs, git_branch: str) -> Outputs:
+    if not git_branch:
+        outputs.run_tests = True
+
+    # Check if we should run the Hypothesis tests
+    if git_branch in {"3.9", "3.10", "3.11"}:
+        print("Branch too old for Hypothesis tests")
+        outputs.run_hypothesis = False
+    else:
+        outputs.run_hypothesis = outputs.run_tests
```

Comment:
At first, I thought that it'd make more sense to detect if there's a hypothesis job:
```suggestion
    # Check if we should run the Hypothesis tests
    if 'test_hypothesis:' not in Path('.github/workflows/build.yml').read_text(encoding='utf-8'):
        print("Branch too old for Hypothesis tests")
        outputs.run_hypothesis = False
    else:
        outputs.run_hypothesis = outputs.run_tests
```

But then, I realized this is still an overkill. This can just be
```suggestion
    # Check if we should run the Hypothesis tests
    outputs.run_hypothesis = outputs.run_tests
```
and that's it. Old branches would just not use this output value so setting it to whatever wouldn't affect anything, no need to set it to `False`.

Reply:
We can simplify further and just replace `run-hypothesis` with `run-tests`

Reply:
Looking into the original conditional, I see that it has the same problem. This value is effectively always the same as `run_tests`. So I'd suggest using `run_tests` in the if-clause of the `run_hypothesis` job. I don't see any reason to have this output separate.

Reply:
> We can simplify further and just replace `run-hypothesis` with `run-tests`

Looks like I posted my comment async :)